operator/v1alpha1/types_image_content_source_policy.go

```diff
@@ -0,0 +1,50 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImageContentSourcePolicy holds cluster-wide information about how to handle registry mirror rules.
+// The canonical name is `cluster`
```

Comment:
There should be no canonical name for this - clients may create multiple of these, it is not a singleton.   We want individual components to be able to register their policies and manage them independently (trying to join a bunch of policies in a doc is very difficult to do with decelerate config models).  That means that we have to define how merging happens, which should be done in the godoc for each field (repositoryDigestMirrors)

Please change the doc to "When multiple policies are defined, the outcome of the behavior is defined on each field".  For the repositoryDigestMirrors field, add a sentence defining how this will work.

For instance, if I have

```
sources:
- a
- b
- c
```

in one doc and 

```
sources:
- c
- d
- e
```

we need to define how they are merged (probably by a zip merge with duplicates removed - `a c b d e`, although we can be loose about the ordering).

Reply:
Keep in mind the use cases we're trying to target when you consider merging:

1. operator on cluster potentially inserts the local registry as a targeted mirror (which probably means it wants to go first)
2. OLM in the future may manage these objects, so it'll be adding them post-install

Reply:
Merged as in when we right these mirror rules to `/etc/containers/registries.conf`?

Reply:
Merged as when they are written and interpreted by all consumers (whether it's registries.conf today, or a future image controller use case like image import).  Documenting the rules of merging makes it clear to consumers and helps clarify the design of components

Reply:
As a counter, does it make sense to disallow having common repos between each config object? That way the administrator can be explicit in how they want the repos ordered?

Reply:
I am thinking of a case where one user of the feature creates an ordering and then a second user comes up with an ordering that has the elements reversed. It might be better to just disallow so they can communicate with first user and come up with a commonly agreed ordering for the repos.

Reply:
I find it a bit attractive to make all of this undefined (explicitly, if necessary).

Let the user declaratively define that some set of repositories is equivalent; let the cluster worry about checking whether they are up, ordering them, possibly load balancing between them, … .  Sure, we do nothing of that kind _now_, but if we define the mirror set to have some kind of strict ordering and strict merging semantics, we’ll lose the opportunity to do something smart in the future (without an explicit opt-in, at least).

I’m especially dubious about the idea of transitively merging mirror sets; sure, it’s trivial to implement, but intuitively it seems to me that if something creates two mirror sets that have a common element but are otherwise distinct (neither a subset of the other), it’s more likely than not that something got very confused and the mirrors are not actually operated to be transitively equivalent that way.

---

Or is the idea that the OpenShift product will ship, and manage, a CRD object that claims the equivalence of `r.a.rh.com/`_foo_ with `r.rh.io/`_bar_ (and OpenShift the organization will commit to keeping the repositories in sync), while the local administrator will set up, and manage, the equivalence of `r.a.rh.com/$foo` with `local-mirror.example.com/`_baz_ (and commit to keeping the two repositories in sync), and therefore we need merging to exist because we don’t want the administrator to edit the OpenShift-product-managed object?

If that is the case, does the administrator need to be able to say “only ever use my mirror without reaching to the internet at all”? That can only work if we define the order of accesses both within a mirror set, _and_ between all the CRD objects (how? based on their name?). (And without an explicit “don’t use those mirrors at all”, the administrator can only put them at the end of the list, to be never contacted unless the local mirror is down. That seems good enough for me, at least in the initial version of this object; `registries.conf` can express “redirect all references to _foo_ to _bar_ without ever contacting _foo_”, but we’ll have enough trouble defining the semantics of merging of the ordering without having to worry about three CRD objects, setting up strict redirects of A→B, B→C, B→A.)

(Or, sure, we could leave the internet-accessible mirrors in and accept them being in the first locations, and just let pulls fail because the DNS or firewall rejects the connection attempts. It could easily be fast enough, but it will cause noise on the firewalls.)

---

(FWIW https://github.com/openshift/machine-config-operator/pull/805 currently sets up the mirrors in `registries.conf` in a rotated fashion, so that different pull specs will result in different order of attempts. That’s just an initial guess, and there’s nothing at all to say that it matters because the pull specs from the different mirrors should not really exist in the cluster at the same time, it can of course be changed to any other semantics.)

Reply:
> Or is the idea that the OpenShift product will ship, and manage, a CRD object that claims the equivalence of r.a.rh.com/foo with r.rh.io/bar (and OpenShift the organization will commit to keeping the repositories in sync),

this is the most likely near term.  The most likely case will be `r.rh.io/*` to `registry.internal.svc/x/y` for digest mirroring of core public resources.

> I’m especially dubious about the idea of transitively merging mirror sets; sure, it’s trivial to implement, but intuitively it seems to me that if something creates two mirror sets that have a common element but are otherwise distinct (neither a subset of the other), it’s more likely than not that something got very confused and the mirrors are not actually operated to be transitively equivalent that way.

Since we're pulling by digest the only expense is check time (which agree can be left as weakly undefined behavior).  If you say A and B are candidates for pull by digest and B and C are candidates by digest and A and B are down I would prefer to check C (this feature is about availability first, performance / local pulls second).

I am not opposed to leaving it weakly undefined, but it needs to satisfy the intent.

As far as order, I would prefer that we preserve preferential rather than round robin, the requirements for this feature are to let you preferentially use a local proxy for performance, and round robin does not satisfy that.